src/praisonai/praisonai/agents_generator.py

```
@@ -407,6 +425,150 @@ def _run_autogen(self, config, topic, tools_dict):
+            for tool_name in details.get("tools", []):
+                tool = tools_dict.get(tool_name)
+                if tool is None:
+                    continue
+                func = tool if callable(tool) else getattr(tool, "run", None)
+                if func is None:
+                    continue
+
+                def make_tool_fn(f):
+                    def tool_fn(**kwargs):
+                        return f(**kwargs) if callable(f) else str(f)
+                    tool_fn.__name__ = tool_name
```

Ruff (0.15.9)

[warning] Function definition does not bind loop variable `tool_name`

(B023)